File: tests/atoms/type-system/apps/chem-types-test.cc
/*
 * chem-types-test.cc
 *
 * Unit test for custom atom types with custom short names.
 * Based on examples/type-system/apps/chemain.cc
 */

#include <stdio.h>
#include <string>
#include <cassert>

#include <opencog/atomspace/AtomSpace.h>

// These include files are in the build directory.
#include "tests/atoms/type-system/demo-types/chem_types.h"
#include "tests/atoms/type-system/demo-types/atom_names.h"

using namespace opencog;

int main(int argc, char* argv[])
{
	printf("Testing custom atom types with short names...\n");

	// Create the AtomSpace into which we'll place everything.
	AtomSpacePtr asp = createAtomSpace();

	// Test 1: Create atoms using short-name constructors
	printf("Test 1: Short-name constructors...\n");
	Handle methane = asp->add_atom(
		Molecule(
			SB(C("carbon"), H("proton 1")),
			SB(C("carbon"), H("proton 2")),
			SB(C("carbon"), H("proton 3")),
			SB(C("carbon"), H("proton 4"))));

	assert(methane != nullptr);
	printf("  Created methane molecule: OK\n");

	// Verify the types are correct
	Handle carbon = asp->add_atom(C("carbon"));
	assert(carbon->get_type() == CARBON_NODE);
	printf("  Carbon type check: OK\n");

	Handle hydrogen = asp->add_atom(H("proton 1"));
	assert(hydrogen->get_type() == HYDROGEN_NODE);
	printf("  Hydrogen type check: OK\n");

	// Test 2: Create atoms using long-form type names
	printf("Test 2: Long-form type constructors...\n");
	Handle water = asp->add_atom(
		createLink(MOLECULE,
			createLink(SINGLE_BOND,
				createNode(OXYGEN_NODE, "oxy"),
				createNode(HYDROGEN_NODE, "h1")),
			createLink(SINGLE_BOND,
				createNode(OXYGEN_NODE, "oxy"),
				createNode(HYDROGEN_NODE, "h2"))));

	assert(water != nullptr);
	printf("  Created water molecule: OK\n");

	// Test 3: Verify type constants are usable
	printf("Test 3: Type constants...\n");
	Type ta = CARBON_NODE;
	Type tb = NITROGEN_NODE;
	std::string sa = "carbo";
	std::string sb = "nitro";

	Handle cyanide = asp->add_atom(
		createLink(MOLECULE,
			createLink(TRIPLE_BOND,
				createNode(ta, sa),
				createNode(tb, sb))));

	assert(cyanide != nullptr);
	printf("  Created cyanide using type constants: OK\n");

	// Test 4: Verify custom types without short names work
	printf("Test 4: Types without custom short names...\n");
	Handle isotope = asp->add_atom(createNode(ISOTOPE_NODE, "some isotope"));
	assert(isotope != nullptr);
	assert(isotope->get_type() == ISOTOPE_NODE);
	printf("  IsotopeNode type check: OK\n");

	// Test 5: Verify all bond types work with short names
	printf("Test 5: All bond types (short names)...\n");
	Handle sb_bond = asp->add_atom(SB(C("c1"), C("c2")));
	Handle db_bond = asp->add_atom(DB(C("c3"), C("c4")));
	Handle tb_bond = asp->add_atom(TB(C("c5"), N("n1")));
	Handle ab_bond = asp->add_atom(AB(C("c6"), C("c7")));

	assert(sb_bond->get_type() == SINGLE_BOND);
	assert(db_bond->get_type() == DOUBLE_BOND);
	assert(tb_bond->get_type() == TRIPLE_BOND);
	assert(ab_bond->get_type() == AROMATIC_BOND);
	printf("  All bond types (short names): OK\n");

	// Test 6: Long names (CamelCase) also work
	printf("Test 6: Long CamelCase names...\n");
	Handle sb_long = asp->add_atom(SingleBond(Carbon("lc1"), Hydrogen("lh1")));
	Handle db_long = asp->add_atom(DoubleBond(Carbon("lc2"), Oxygen("lo1")));
	Handle tb_long = asp->add_atom(TripleBond(Carbon("lc3"), Nitrogen("ln1")));

	assert(sb_long->get_type() == SINGLE_BOND);
	assert(db_long->get_type() == DOUBLE_BOND);
	assert(tb_long->get_type() == TRIPLE_BOND);
	printf("  Long CamelCase names: OK\n");

	printf("\nAll tests passed!\n");
	return 0;
}

File: tests/atoms/type-system/apps/chem-types-test.py
#!/usr/bin/env python3
#
# chem-types-test.py
#
# Unit test for custom atom types with custom short names in Python.
# Based on examples/type-system/apps/chemy_hello.py
#

import unittest
import sys

# Import the AtomSpace, and the basic AtomSpace types
from opencog.atomspace import AtomSpace
from opencog.atomspace import set_thread_atomspace

# Import all of the chemical element types, and bond types too.
from test_chempydemo import *


class ChemTypesTest(unittest.TestCase):

    def setUp(self):
        self.spa = AtomSpace()
        set_thread_atomspace(self.spa)

    def tearDown(self):
        pass

    def test_create_atoms_with_short_names(self):
        """Test creating atoms using short-name constructors"""
        # Create a Magnesium atom
        mg = Mg('foo')
        self.assertIsNotNone(mg)
        # Custom-named types use their short name as TYPE_NAME
        self.assertEqual(mg.type_name, 'Mg')

    def test_create_bonds(self):
        """Test creating bonds with short-name constructors"""
        ch_bond = SB(C('carbon'), H('hydrogen'))
        self.assertIsNotNone(ch_bond)
        # Custom-named types use their short name as TYPE_NAME
        self.assertEqual(ch_bond.type_name, 'SB')

    def test_create_molecule(self):
        """Test creating a molecule"""
        methane = Molecule(
            SB(C('1'), H('1')),
            SB(C('1'), H('2')),
            SB(C('1'), H('3')),
            SB(C('1'), H('4')))
        self.assertIsNotNone(methane)
        self.assertEqual(methane.type_name, 'Molecule')

    def test_element_types(self):
        """Test various element types"""
        hydrogen = H('h1')
        carbon = C('c1')
        nitrogen = N('n1')
        oxygen = O('o1')

        # Custom-named types use their short name as TYPE_NAME
        self.assertEqual(hydrogen.type_name, 'H')
        self.assertEqual(carbon.type_name, 'C')
        self.assertEqual(nitrogen.type_name, 'N')
        self.assertEqual(oxygen.type_name, 'O')

    def test_bond_types(self):
        """Test all bond types"""
        single = SB(C('c1'), C('c2'))
        double = DB(C('c3'), C('c4'))
        triple = TB(C('c5'), N('n1'))
        aromatic = AB(C('c6'), C('c7'))

        # Custom-named types use their short name as TYPE_NAME
        self.assertEqual(single.type_name, 'SB')
        self.assertEqual(double.type_name, 'DB')
        self.assertEqual(triple.type_name, 'TB')
        self.assertEqual(aromatic.type_name, 'AB')

    def test_carbon14_execution(self):
        """Test Carbon14 atom execution"""
        c14 = Carbon14Node('found in wood')
        self.assertIsNotNone(c14)
        self.assertEqual(c14.type_name, 'Carbon14Node')

        decay_products = c14.execute()
        self.assertIsNotNone(decay_products)

        # Check that decay products contain expected types
        products_list = decay_products.to_list()
        self.assertTrue(len(products_list) > 0)

        # First product should be Nitrogen14Node
        nitrogen = products_list[0]
        self.assertEqual(nitrogen.type_name, 'Nitrogen14Node')


if __name__ == '__main__':
    unittest.main()
